include/tvm/runtime/packed_func.h

```diff
@@ -2129,6 +2315,42 @@ struct PackedFuncValueConverter<::tvm::runtime::String> {
   }
 };
 
+template <typename T>
+struct PackedFuncValueConverter<Array<T>> {
```

Comment:
I feel automatic conversion here is a bit overkill.  Recursive automatic conversion would be too much overhead whe passing through FFI. The developer should ensure the correct typings of the internal contained values

I know it may bring some convenience, but in this case I think we should not do it. 

Suggestion: Remove the conversion logic from this PR

Reply:
I think we need these conversions in order to preserve backwards compatibility with existing user-written code.  Otherwise, a user-written function that accepts `Array<PrimExpr>` could no longer be called with a python list `[1,2,3]`.

Reply:
Regarding performance, I think we should be good there as well.  The conversion is applied instead of the existing recursive check in `ObjectTypeChecker<Array<T>>`, so no additional recursive walks.  The mutation is implemented in terms of `Array::Map`, which is copy-on-write, so no allocations occur if the user has provided the expected types (the common case).

Reply:
We should likely check some of those cases, and update quite a few of them to `Array<runtime::Int>`. But agree some might be useful especially for the mixed symbolic shape  and static shape case

Reply:
added some comments below on optimizing for freq case

Reply:
Sounds good, and going through the comments now.  I like the idea of updating the call sites over time, so that eventually the backwards compatibility conversion would only be triggered for externally-defined PackedFunc instances.